src/twisted/mail/test/test_smtp.py

```diff
@@ -674,6 +676,50 @@ def check(ignored):
 
         return self.loopback(server, client).addCallback(check)
 
+    def test_ESMTPSenderFactory_TLSError(self):
+        """
+        Attempting to connect to an ESMTP server which presents an invalid certificate
+        should produce a meaningful error.
+        """
+        # set up a dummy ESMTP server which will present a self-signed cert after
+        # STARTTLS
+        buildServerProtocol = lambda: DummyESMTP(contextFactory=ServerTLSContext())
+        serverFactory = Factory.forProtocol(buildServerProtocol)
+        serverPort = reactor.listenTCP(0, serverFactory, interface="127.0.0.1")
+        self.addCleanup(serverPort.stopListening)
+
+        # build a client, which won't accept the certificate presented by the dummy
+        # server.
+        sentDeferred = defer.Deferred()
+        clientFactory = smtp.ESMTPSenderFactory(
+            "username",
+            "password",
+            "source@address",
+            "recipient@address",
+            BytesIO(b"message"),
+            sentDeferred,
+            retries=0,
+            requireAuthentication=False,
+            contextFactory=optionsForClientTLS("testdomain"),
+        )
+
+        # connect the two together
+        connector = reactor.connectTCP(
+            serverPort.getHost().host, serverPort.getHost().port, clientFactory
+        )
+        self.addCleanup(connector.disconnect)
+
+        def errback(f):
+            # check that we got an exception with a plausible error code.
+            f.trap(smtp.SMTPConnectError)
+            self.assertIn("certificate verify failed", str(f.value))
+
+        def callback(_res):
+            raise Exception("ESMTPSenderFactory unexpectedly completed")
+
+        sentDeferred.addCallbacks(callback, errback)
+        return sentDeferred
```

Comment:
In general we now strongly prefer the style of test where you *don't* spin up a real reactor and do for-real network I/O; instead, consider using the various utilities in `twisted.internet.testing` to produce deterministic mocks for your network set-up attempts.  That said, this style of test is definitely better than nothing, so if you feel that the deterministic version would take you a bunch more time we can work with this for now (it's certainly not the first such test in this file).

Reply:
Thanks for the speedy initial review, @glyph - and sorry for the slow turnaround on my side.

Yeah, I'm aware of the preference. I initially just went with the flow in this file, but I've just taken a bit of a harder look at what it would take to use `twisted.test.iosim.FakeTransport` and friends to connect up the client and server.

It would be a bit fiddly, but the real problem is that what I really want to test here is how `_newtls.startTLS` interacts with the SMTP code, and `startTLS` relies on wrapping a real `FileDescriptor` object - or at least, something that behaves much the same as one. Possibly I could rejig `FakeTransport` enough to make `startTLS` work with it ... but at that point I start to feel like I'm just testing the tests.

So, as painful as it is, I think this might be a case where we're better off with a real reactor, at least for now.

Reply:
No need to be sorry for a slow turnaround. We're all very time-constrained here in open source land, and we do things when we can.